File: funcs/sumImageWithinCircle.py
def sumImageWithinCircle(img, xmin, xmax, ymin, ymax, r, cx = 0.0, cy = 0.0, ndiv = 16):
    """
    Sum the pixel values on an image that are within a hard circular mask.

    Arguments:
    img -- 2D image with axes (ny, nx)
    xmin -- left edge of leftmost pixel
    xmax -- right edge of rightmost pixel
    ymin -- lower edge of lowermost pixel
    ymax -- upper edge of uppermost pixel
    r -- radius of circle

    Keyword arguments:
    cx -- x position of centre of circle (default 0.0)
    cy -- y position of centre of circle (default 0.0)
    ndiv -- number sub-divisions (default 16)

    Note:
    This function is crying out for FORTRAN+OpenMP.
    """

    # Import special modules ...
    try:
        import numpy
    except:
        raise Exception("run \"pip install --user numpy\"")

    # Import sub-functions ...
    from .findFractionOfPixelWithinCircle import findFractionOfPixelWithinCircle

    # Create axes relative to the centre of the circle ...
    xaxis = numpy.linspace(xmin, xmax, num = img.shape[1] + 1) - cx
    yaxis = numpy.linspace(ymin, ymax, num = img.shape[0] + 1) - cy

    # Find out the distance of each corner to the centre of the circle ...
    dist = numpy.zeros((yaxis.size, xaxis.size), dtype = numpy.float64)
    for ix in range(xaxis.size):
        for iy in range(yaxis.size):
            dist[iy, ix] = numpy.hypot(xaxis[ix], yaxis[iy])

    # Initialize total ...
    tot = 0.0

    # Loop over x-axis ...
    for ix in range(img.shape[1]):
        # Loop over y-axis ...
        for iy in range(img.shape[0]):
            # Skip this pixel if it is empty ...
            if img[iy, ix] == 0.0:
                continue

            # Skip this pixel if it is all outside of the circle ...
            if numpy.all(dist[iy:iy + 2, ix:ix + 2] >= r):
                continue

            # Check if this pixel is entirely within the circle or if it
            # straddles the circumference ...
            if numpy.all(dist[iy:iy + 2, ix:ix + 2] <= r):
                # Add all of the value to total ...
                tot += img[iy, ix]
            else:
                # Add part of the value to total ...
                tot += img[iy, ix] * findFractionOfPixelWithinCircle(xaxis[ix], xaxis[ix + 1], yaxis[iy], yaxis[iy + 1], r, ndiv = ndiv)

    # Return answer ...
    return tot

File: f90/src/sumImageWithinCircle.f90
PURE SUBROUTINE sumImageWithinCircle(ndiv, nx, ny, xmin, xmax, ymin, ymax, r, cx, cy, img, tot)
    ! Import standard modules ...
    USE ISO_C_BINDING

    IMPLICIT NONE

    ! Declare inputs/outputs ...
    INTEGER(kind = C_LONG_LONG), INTENT(in)                                     :: ndiv
    INTEGER(kind = C_LONG_LONG), INTENT(in)                                     :: nx
    INTEGER(kind = C_LONG_LONG), INTENT(in)                                     :: ny
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: xmin
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: xmax
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: ymin
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: ymax
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: r
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: cx
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: cy
    REAL(kind = C_DOUBLE), DIMENSION(ny, nx), INTENT(in)                        :: img
    REAL(kind = C_DOUBLE), INTENT(out)                                          :: tot

    ! Declare internal variables ...
    INTEGER(kind = C_LONG_LONG)                                                 :: ix
    INTEGER(kind = C_LONG_LONG)                                                 :: iy
    REAL(kind = C_DOUBLE)                                                       :: dx
    REAL(kind = C_DOUBLE)                                                       :: dy
    REAL(kind = C_DOUBLE)                                                       :: frac
    REAL(kind = C_DOUBLE), DIMENSION(nx + 1)                                    :: xaxis
    REAL(kind = C_DOUBLE), DIMENSION(ny + 1)                                    :: yaxis
    REAL(kind = C_DOUBLE), DIMENSION(ny + 1, nx + 1)                            :: dist

    ! Calculate size of pixels ...
    dx = (xmax - xmin) / REAL(ndiv, kind = C_DOUBLE)
    dy = (ymax - ymin) / REAL(ndiv, kind = C_DOUBLE)

    ! Create nodes relative to the centre of the circle ...
    DO ix = 1_C_LONG_LONG, nx + 1_C_LONG_LONG
        xaxis(ix) = xmin + REAL(ix - 1_C_LONG_LONG, kind = C_DOUBLE) * dx - cx
    END DO

    ! Create nodes relative to the centre of the circle ...
    DO iy = 1_C_LONG_LONG, ny + 1_C_LONG_LONG
        yaxis(iy) = ymin + REAL(iy - 1_C_LONG_LONG, kind = C_DOUBLE) * dy - cy
    END DO

    ! Find out the distance of each node to the centre of the circle ...
    DO ix = 1_C_LONG_LONG, nx + 1_C_LONG_LONG
        DO iy = 1_C_LONG_LONG, ny + 1_C_LONG_LONG
            dist(iy, ix) = HYPOT(xaxis(ix), yaxis(iy))
        END DO
    END DO

    ! Initialize total ...
    tot = 0.0e0_C_DOUBLE

    ! Loop over x-axis ...
    DO ix = 1_C_LONG_LONG, nx
        ! Loop over y-axis ...
        DO iy = 1_C_LONG_LONG, ny
            ! Skip this pixel if it is empty ...
            IF(img(iy, ix) == 0.0e0_C_DOUBLE)THEN
                CYCLE
            END IF

            ! Skip this pixel if it is all outside of the circle ...
            IF(ALL(dist(iy:iy + 1_C_LONG_LONG, ix:ix + 1_C_LONG_LONG) >= r))THEN
                CYCLE
            END IF

            ! Check if this pixel is entirely within the circle or if it
            ! straddles the circumference ...
            IF(ALL(dist(iy:iy + 1_C_LONG_LONG, ix:ix + 1_C_LONG_LONG) <= r))THEN
                ! Add all of the value to total ...
                tot = tot + img(iy, ix)
            ELSE
                ! Add part of the value to total ...
                CALL findFractionOfPixelWithinCircle(                           &
                    ndiv = ndiv,                                                &
                    xmin = xaxis(ix),                                           &
                    xmax = xaxis(ix + 1_C_LONG_LONG),                           &
                    ymin = yaxis(iy),                                           &
                    ymax = yaxis(iy + 1_C_LONG_LONG),                           &
                    r = r,                                                      &
                    cx = 0.0e0_C_DOUBLE,                                        &
                    cy = 0.0e0_C_DOUBLE,                                        &
                    frac = frac                                                 &
                )
                tot = tot + img(iy, ix) * frac
            END IF
        END DO
    END DO
END SUBROUTINE sumImageWithinCircle
